### PR TITLE
Council motions

### PR DESCRIPTION
**Pull request by kwingram25**
Changes treasury proposals to rely on council motions as designed and adding new interface to view, create, etc.

Not finished as councilMotions.voting is currently unusable - waiting on api work

### CODE REVIEW

File: packages/app-council/src/Motions/Motion.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
+// Copyright 2017-2019 @polkadot/app-democracy authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { I18nProps } from '@polkadot/ui-app/types';
+import { RawParam } from '@polkadot/ui-params/types';
+import { Option, Proposal, Votes } from '@polkadot/types';
```
Comment (Tbaut) [inline packages/app-council/src/Motions/Motion.tsx]: (nit) alphabetical order, same for the imports below

File: packages/app-council/src/Motions/Motion.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+const COLORS_YAY = settings.uiTheme === 'substrate'
+  ? ['#4d4', '#4e4']
+  : ['#64bebe', '#5badad'];
+const COLORS_NAY = settings.uiTheme === 'substrate'
+  ? ['#d44', '#e44']
+  : ['#d75ea1', '#e189ba'];
+
+type Props = I18nProps & {
+  hash: string,
```
Comment (Tbaut) [inline packages/app-council/src/Motions/Motion.tsx]: (nit) alphabetical order

File: packages/app-council/src/Motions/index.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2017-2019 @polkadot/app-democracy authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { I18nProps } from '@polkadot/ui-app/types';
+
+import React from 'react';
+import { Hash } from '@polkadot/types';
```
Comment (Tbaut) [inline packages/app-council/src/Motions/index.tsx]: (nit) this is a type and should be in the group above + alphabetical order :)

File: packages/app-treasury/src/Overview/Approve.tsx [hunk 4]
```diff
@@ -1,19 +1,22 @@
 // Copyright 2017-2019 @polkadot/app-democracy authors & contributors
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
-
```
Comment (Tbaut) [inline packages/app-treasury/src/Overview/Approve.tsx]: you can keep this empty line between the header and the imports

File: packages/app-treasury/src/Overview/Approve.tsx [hunk 5]
```diff
@@ -33,8 +36,33 @@ class Approve extends TxModal<Props, State> {
 
   headerText = () => this.props.t('Approve or reject proposal');
 
-  txMethod = () => this.state.isApproving ? 'treasury.approveProposal' : 'treasury.rejectProposal';
-  txParams = () => [this.props.proposalId];
+  txMethod = () => 'councilMotions.propose';
+  txParams = () => {
+    const { api, proposalId, threshold } = this.props;
+    const { isApproving } = this.state;
+
```
Comment (Tbaut) [inline packages/app-treasury/src/Overview/Approve.tsx]: (nit) I guess we can remove this empty line and the one 2 lines after as well

File: packages/ui-app/src/Collection.tsx [hunk 6]
```diff
@@ -43,12 +43,15 @@ export default class Collection<P extends CollectionProps, S extends CollectionS
...
     this.state = {
-      isEmpty: props.children && (props.children as Array<any>).length > 0
+      isEmpty: props.isEmpty !== undefined ? props.isEmpty : props.children && (props.children as Array<any>).length > 0
     } as S;
   }
 
-  static getDerivedStateFromProps ({ children }: CollectionProps) {
-    if (!children || (children as Array<any>).length <= 0) {
+  static getDerivedStateFromProps ({ children, isEmpty }: CollectionProps) {
+    if (isEmpty !== undefined) {
```
Comment (Tbaut) [inline packages/ui-app/src/Collection.tsx]: I guess this can be simplified with `if (!!isEmpty)`

File: packages/ui-app/src/TxModal.tsx [hunk 7]
```diff
@@ -17,44 +16,50 @@ export type TxModalProps = I18nProps & {
...
   } as S;
 
   state: S = this.defaultState;
 
-  componentWillReceiveProps ({ isOpen }: P, _: S) {
-    if (isOpen && !this.props.isOpen && !this.state.isBusy) {
-      this.reset();
-    }
-  }
+  // componentWillReceiveProps ({ isOpen }: P, _: S) {
```
Comment (Tbaut) [inline packages/ui-app/src/TxModal.tsx]: (nit) stray comments

File: packages/ui-app/src/TxModal.tsx [hunk 8]
```diff
@@ -111,6 +110,8 @@ class TxModal<P extends TxModalProps, S extends TxModalState> extends TxComponen
   protected txMethod: () => string = () => '';
   protected txParams: () => Array<any> = () => [];
 
+  protected renderTrigger?: () => React.ReactNode | null = () => null;
+
```
Comment (Tbaut) [inline packages/ui-app/src/TxModal.tsx]: (nit) no need for this empty line right?

File: packages/app-council/src/Motions/Motion.tsx [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+import styled from 'styled-components';
+import { ActionItem, Chart, Voting } from '@polkadot/ui-app';
+import VoteThreshold from '@polkadot/ui-params/Param/VoteThreshold';
+import { withCalls, withMulti } from '@polkadot/ui-api';
+import settings from '@polkadot/ui-settings';
+import { formatNumber } from '@polkadot/util';
+
+import translate from '../translate';
+
+const COLORS_YAY = settings.uiTheme === 'substrate'
```
Comment (jacogr) [inline packages/app-council/src/Motions/Motion.tsx]: Not crazy about this duplication at all. We should rather have it external and shared between the apps using it.

**Comment by kwingram25** [comment]
Added voting on council motions and more in-depth views for voting in general - better to see what is proposed while casting a vote

Several treasury api methods have disappeared, including pot and all the settings so Edit Settings no longer works

**Comment by jacogr** [comment]
I would just remove edit settings completely. I'm assuming all those were moved to constants. (We can re-visit what needs to go as summary and what we can set once things stabilizes)

**Comment by jacogr** [comment]
For me, I see 3 proposals. They are all empty - is this expected on a dev chain?

image

Additionally, move the submit button to top-level. (Like we have elsewhere)

Then when trying to approve, seeing this -

image

I don't think it quite fits with the L&F we have elsewhere, actually looks out of place. It feels like a repeat of info here as well, we already have it on-screen. I understand that that is the one we are busy with, however then rather shorter/summary if really needed,. (Or just the number which we have anyway)

The above needs some thought, i.e. how do we have short-to-the-point summary info for all of these voting modals.

File: packages/app-council/src/Motions/Motion.tsx [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  chain_bestNumber?: BN,
+  hash: string,
+  proposal: Proposal | null,
+  votes: Votes | null
+};
+
+type State = {
+  votedTotal: number,
+  votedNay: number,
+  votedYay: number
```
Comment (Tbaut) [inline packages/app-council/src/Motions/Motion.tsx]:
```suggestion
  votedAye: number
```

File: packages/app-council/src/Motions/Motion.tsx [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+class Motion extends React.PureComponent<Props, State> {
+  state: State = {
+    votedTotal: 0,
+    votedYay: 0,
```
Comment (Tbaut) [inline packages/app-council/src/Motions/Motion.tsx]:
```suggestion
    votedAye: 0,
```

File: packages/app-council/src/Motions/Motion.tsx [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+  static getDerivedStateFromProps ({ votes }: Props): State | null {
+    if (!votes) {
+      return null;
+    }
+
+    const { ayes, nays } = votes;
+
+    const newState: State = {
+      votedNay: nays.length,
+      votedYay: ayes.length,
```
Comment (Tbaut) [inline packages/app-council/src/Motions/Motion.tsx]:
```suggestion
      votedAye: ayes.length,
```

File: packages/ui-app/src/Voting.tsx [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+  constructor (props: Props) {
+    super(props);
+
+    const { t } = props;
+
+    this.state = {
+      ...this.defaultState,
+      voteOptions: [
+        { text: t('Yay, I approve'), value: true },
```
Comment (Tbaut) [inline packages/ui-app/src/Voting.tsx]:
```suggestion
        { text: t('Aye, I approve'), value: true },
```

File: packages/ui-app/src/Voting.tsx [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  state: State;
+
+  headerText = () => {
+    const { isCouncil, t } = this.props;
+
+    return t(isCouncil ? 'Vote on council motion' : 'Vote on proposal');
+  }
+
+  accountLabel = () => this.props.t('Vote with account');
+  accountHelp = () => this.props.t('Select the account you wish to vote with. You can approve "yay" or deny "nay" the proposal.');
```
Comment (Tbaut) [inline packages/ui-app/src/Voting.tsx]:
```suggestion
  accountHelp = () => this.props.t('Select the account you wish to vote with. You can approve "aye" or deny "nay" the proposal.');
```

**Comment by kwingram25** [comment]
- Made voting view of proposals collapsed by default - open to ideas how to make it look less out of place
- Changed "yay" to "aye"

**Pull request merged**